File: app/src/main/java/org/egon12/btmid/midi/MidiParser.kt
package org.egon12.btmid.midi

object MidiParser {

    fun parse(msg: ByteArray, offset: Int, count: Int): List<MidiEvent> {
        val events = mutableListOf<MidiEvent>()
        var i = offset
        val end = offset + count
        var runningStatus = 0

        while (i < end) {
            val b = msg[i].u()
            val status: Int
            val d: Int  // index of first data byte

            if (b and 0x80 != 0) {
                status = b
                runningStatus = b
                d = i + 1
            } else {
                status = runningStatus
                d = i
            }

            val type = status and 0xF0
            val ch = status and 0x0F

            when (type) {
                0x80 -> {
                    if (d + 1 >= end) break
                    events += MidiEvent.NoteOff(ch, msg[d].u(), msg[d + 1].u())
                    i = d + 2
                }
                0x90 -> {
                    if (d + 1 >= end) break
                    val note = msg[d].u()
                    val vel = msg[d + 1].u()
                    events += if (vel == 0) MidiEvent.NoteOff(ch, note, 0)
                              else MidiEvent.NoteOn(ch, note, vel)
                    i = d + 2
                }
                0xB0 -> {
                    if (d + 1 >= end) break
                    events += MidiEvent.ControlChange(ch, msg[d].u(), msg[d + 1].u())
                    i = d + 2
                }
                0xA0, 0xE0 -> i = d + 2  // skip other 2-data-byte messages
                0xC0, 0xD0 -> i = d + 1  // skip 1-data-byte messages
                else        -> i = d + 1  // system/unknown — advance one byte
            }
        }
        return events
    }

    private fun Byte.u() = toInt() and 0xFF
}

File: app/src/main/cpp/MidiParser.cpp
#include "MidiParser.h"

int parseMidi(const uint8_t* data, size_t length, MidiMsg* out, int maxOut, uint8_t& runningStatus) {
    int    count = 0;
    size_t i     = 0;

    while (i < length && count < maxOut) {
        uint8_t b = data[i];
        uint8_t status;
        size_t  d;  // index of first data byte

        if (b & 0x80) {
            status        = b;
            runningStatus = b;
            d             = i + 1;
        } else {
            status = runningStatus;
            d      = i;
        }

        uint8_t type    = status & 0xF0;
        uint8_t channel = status & 0x0F;

        bool stop = false;
        switch (type) {
            case 0x80:
                if (d + 2 > length) { stop = true; break; }
                out[count++] = { MidiMsgType::NoteOff, channel, data[d], data[d + 1] };
                i = d + 2;
                break;
            case 0x90: {
                if (d + 2 > length) { stop = true; break; }
                uint8_t vel = data[d + 1];
                out[count++] = { vel == 0 ? MidiMsgType::NoteOff : MidiMsgType::NoteOn,
                                 channel, data[d], vel };
                i = d + 2;
                break;
            }
            case 0xB0:
                if (d + 2 > length) { stop = true; break; }
                out[count++] = { MidiMsgType::CC, channel, data[d], data[d + 1] };
                i = d + 2;
                break;
            case 0xA0: case 0xE0: i = d + 2; break;
            case 0xC0: case 0xD0: i = d + 1; break;
            default:               i = d + 1; break;
        }
        if (stop) break;
    }
    return count;
}
